### PR TITLE
Quadrotor Simulation in C++

### PR DESCRIPTION
**Pull request by psiorx**
As discussed in the meeting, this currently uses the ode1 pipeline.  At a timestep of 0.005s it seems to be fine.  I'll bring in a more robust solver to the C++ simulation capabilities in a second pass.

Currently this has two example executables:

* One which grabs inputs over LCM and publishes state (will use for FLA).
* An LQR hovering example showing closed loop feedback control.

Performance seems great so far since it can simulate at the target rate of 200 Hz while consuming less than 3% of a CPU core.  Simulink couldn't keep up with realtime even at 100 Hz.

### CODE REVIEW

File: drake/examples/Quadrotor/Quadrotor.h [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    Eigen::Matrix<Scalar, 3, 1> rpy_ddot = Phi * R * pqr_dot + dPhi_x_rpydot * R * pqr + Phi * Rdot * pqr;
+
+    xdot << x.template tail<6>(), xyz_ddot, rpy_ddot;
+
+    return xdot;
+  }
+
+};
+
+class QuadrotorWithBotVis : public Quadrotor {
```
Comment (tkoolen) [inline drake/examples/Quadrotor/Quadrotor.h]: I know this is pattern matched from Pendulum, but I really don't like using inheritance for this. It's like:
Animal < Cat < BeautifulCat

File: drake/examples/Quadrotor/Quadrotor.h [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    angularvel2rpydotMatrix(rpy, Phi, &dPhi, ddPhi);
+
+    Eigen::Matrix<Scalar, 9, 3> drpy2drotmat = drpy2rotmat(rpy);
+    Eigen::Matrix<Scalar, 9, 1> Rdot_vec;
+    Rdot_vec = drpy2drotmat * rpydot;
+    Eigen::Matrix<Scalar, 3, 3> Rdot = Eigen::Map< Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic> >(Rdot_vec.data(), 3, 3);
+    Eigen::Matrix<Scalar, 9, 1> dPhi_x_rpydot_vec;
+    dPhi_x_rpydot_vec = dPhi * rpydot;
+    Eigen::Matrix<Scalar, 3, 3> dPhi_x_rpydot = Eigen::Map< Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic> >(dPhi_x_rpydot_vec.data(), 3, 3);
+    Eigen::Matrix<Scalar, 3, 1> rpy_ddot = Phi * R * pqr_dot + dPhi_x_rpydot * R * pqr + Phi * Rdot * pqr;
```
Comment (tkoolen) [inline drake/examples/Quadrotor/Quadrotor.h]: This is fine, but we're really just doing differentiation by hand again. Why not use AutoDiff? :-) 

To compute `Rdot`, you can create an AutoDiffScalar vector `rpy_autodiff`, which has `rpy` as values and `rpydot` as derivatives (i.e. derivatives only w.r.t. time). Then `rpy2rotmat(rpy_autodiff)` will compute both `R_autodiff`, which has both `R` and `Rdot` in it.

Next, to compute `pqr_dot`, you can similarly create an AutoDiffScalar vector `pqr_autodiff`, which has `pqr` as values and `pqr_dot` as derivatives. Then you just take the algorithm for computing `rpy_dot` from `rpy`, `pqr` and `R`, i.e.
```
rpy_dot = angularvel2rpydotMatrix(rpy) * R * pqr;
```
and call it with the AutoDiff versions (`rpy_autodiff`, `pqr_autodiff` and `R_autodiff`), and you'll get `rpy_ddot` as the gradient output.

I implemented this here:
https://github.com/tkoolen/crtpDemo/blob/master/example/QuadPlantPenn.h

**Comment by psiorx** [comment]
Thanks for the comments.  I agree with both of them.  I'll probably incorporate the changes in a later PR since the code underlying this is getting overhauled soon.

**Comment by tkoolen** [comment]
That's fine of course.

File: drake/examples/Quadrotor/Quadrotor.h (hunk 1 above)
Comment (RussTedrake) [inline drake/examples/Quadrotor/Quadrotor.h]: again, this is just a short-term hack until we can properly cascade the visualizer with systems.

**Pull request merged**